File: apps/web/src/types/dataset.ts
export type DatasetStatus = "uploading" | "processing" | "ready" | "failed" | "archived";
export type ColumnType = "integer" | "float" | "string" | "boolean" | "datetime" | "date" | "unknown";

export interface ColumnProfile {
  name: string;
  dtype: ColumnType;
  null_count: number;
  null_pct: number;
  unique_count: number;
  cardinality: number;
  sample_values: unknown[];
  min_value: number | string | null;
  max_value: number | string | null;
  mean_value: number | null;
  std_value: number | null;
  skewness: number | null;
  kurtosis: number | null;
  quantiles: Record<string, number>;
  top_values: [unknown, number][];
}

export interface DatasetSchema {
  columns: ColumnProfile[];
  row_count: number;
  column_count: number;
  size_bytes: number;
  inferred_primary_key: string | null;
  inferred_time_column: string | null;
  inferred_target_column: string | null;
  correlation_matrix: Record<string, Record<string, number>>;
  data_quality_score: number;
}

export interface Dataset {
  id: string;
  name: string;
  description: string | null;
  original_filename: string;
  file_format: string;
  status: DatasetStatus;
  row_count: number | null;
  column_count: number | null;
  data_quality_score: number | null;
  analysis_count: number;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface DatasetDetail extends Dataset {
  schema: DatasetSchema | null;
  metadata: Record<string, unknown>;
}

export interface DatasetUploadParams {
  file: File;
  name: string;
  description?: string;
  tags?: string[];
}

export type AnalysisType =
  | "eda"
  | "anomaly_detection"
  | "forecasting"
  | "clustering"
  | "classification"
  | "regression"
  | "correlation"
  | "custom_query"
  | "full_pipeline";

export type AnalysisStatus = "queued" | "running" | "completed" | "failed" | "cancelled";

export interface AgentStep {
  step_number: number;
  agent_name: string;
  action: string;
  tool_name: string | null;
  tool_input: Record<string, unknown>;
  tool_output: string | null;
  reasoning: string;
  sql_executions: SQLExecution[];
  duration_ms: number;
  timestamp: string;
}

export interface SQLExecution {
  id: string;
  query: string;
  natural_language_intent: string | null;
  execution_time_ms: number;
  row_count: number;
  result_preview: Record<string, unknown>[];
  execution_plan: string | null;
  error: string | null;
  executed_at: string;
}

export interface ForecastPoint {
  timestamp: string;
  value: number;
  lower_bound: number;
  upper_bound: number;
  is_forecast: boolean;
}

export interface AnomalyRecord {
  row_index: number;
  anomaly_score: number;
  algorithm: string;
  affected_columns: string[];
  explanation: string;
  severity: "low" | "medium" | "high" | "critical";
}

export interface MLResult {
  model_name: string;
  task_type: string;
  metrics: Record<string, number>;
  feature_importance: Record<string, number>;
  shap_values: Record<string, number[]>;
  forecast_points: ForecastPoint[];
  anomalies: AnomalyRecord[];
  natural_language_summary: string;
}

export interface AnalysisResult {
  summary: string;
  key_findings: string[];
  sql_executions: SQLExecution[];
  ml_results: MLResult[];
  visualizations: unknown[];
  confidence_score: number;
  evidence_references: string[];
}

export interface Analysis {
  id: string;
  dataset_id: string;
  analysis_type: AnalysisType;
  status: AnalysisStatus;
  user_query: string | null;
  duration_seconds: number | null;
  step_count: number;
  created_at: string;
  completed_at: string | null;
}

export interface AnalysisDetail extends Analysis {
  agent_steps: AgentStep[];
  result: AnalysisResult | null;
  error_message: string | null;
}

export type InsightCategory =
  | "anomaly" | "trend" | "correlation" | "forecast"
  | "recommendation" | "data_quality" | "business" | "statistical";

export type InsightSeverity = "info" | "warning" | "critical" | "opportunity";

export interface Insight {
  id: string;
  dataset_id: string;
  analysis_id: string | null;
  title: string;
  body: string;
  category: InsightCategory;
  severity: InsightSeverity;
  confidence_score: number;
  affected_columns: string[];
  tags: string[];
  business_impact: string | null;
  recommended_actions: string[];
  is_verified: boolean;
  created_at: string;
}

File: apps/api/src/domain/entities/dataset.py
from __future__ import annotations

import uuid
from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field, field_validator


class DatasetStatus(StrEnum):
    UPLOADING = "uploading"
    PROCESSING = "processing"
    READY = "ready"
    FAILED = "failed"
    ARCHIVED = "archived"


class ColumnType(StrEnum):
    INTEGER = "integer"
    FLOAT = "float"
    STRING = "string"
    BOOLEAN = "boolean"
    DATETIME = "datetime"
    DATE = "date"
    UNKNOWN = "unknown"


class ColumnProfile(BaseModel):
    name: str
    dtype: ColumnType
    null_count: int
    null_pct: float
    unique_count: int
    cardinality: float  # unique_count / total_rows
    sample_values: list[Any] = Field(default_factory=list, max_length=10)
    min_value: float | str | None = None
    max_value: float | str | None = None
    mean_value: float | None = None
    std_value: float | None = None
    skewness: float | None = None
    kurtosis: float | None = None
    quantiles: dict[str, float] = Field(default_factory=dict)
    top_values: list[tuple[Any, int]] = Field(default_factory=list)


class DatasetSchema(BaseModel):
    columns: list[ColumnProfile]
    row_count: int
    column_count: int
    size_bytes: int
    inferred_primary_key: str | None = None
    inferred_time_column: str | None = None
    inferred_target_column: str | None = None
    correlation_matrix: dict[str, dict[str, float]] = Field(default_factory=dict)
    data_quality_score: float = Field(ge=0.0, le=1.0, default=0.0)


class DatasetVersion(BaseModel):
    id: uuid.UUID = Field(default_factory=uuid.uuid4)
    dataset_id: uuid.UUID
    version_number: int
    file_path: str
    schema_snapshot: DatasetSchema | None = None
    row_count: int
    size_bytes: int
    created_at: datetime = Field(default_factory=datetime.utcnow)
    notes: str | None = None


class Dataset(BaseModel):
    id: uuid.UUID = Field(default_factory=uuid.uuid4)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    original_filename: str
    file_path: str
    file_format: str  # csv | parquet | json
    status: DatasetStatus = DatasetStatus.UPLOADING
    schema: DatasetSchema | None = None  # type: ignore[assignment]
    current_version: int = 1
    versions: list[DatasetVersion] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    analysis_count: int = 0

    @field_validator("file_format")
    @classmethod
    def validate_format(cls, v: str) -> str:
        allowed = {"csv", "parquet", "json", "jsonl"}
        if v.lower() not in allowed:
            raise ValueError(f"Unsupported format '{v}'. Allowed: {allowed}")
        return v.lower()

    def mark_ready(self, schema: DatasetSchema) -> None:
        self.schema = schema
        self.status = DatasetStatus.READY
        self.updated_at = datetime.utcnow()

    def mark_failed(self) -> None:
        self.status = DatasetStatus.FAILED
        self.updated_at = datetime.utcnow()

    def increment_analysis_count(self) -> None:
        self.analysis_count += 1
        self.updated_at = datetime.utcnow()
